security.go

```diff
@@ -1,47 +1,91 @@
-package slack
-
-import (
-	"crypto/hmac"
-	"crypto/sha256"
-	"encoding/hex"
-	"errors"
-	"fmt"
-	"hash"
-	"net/http"
-)
-
-// SecretsVerifier contains the information needed to verify that the request comes from Slack
-type SecretsVerifier struct {
-	slackSig  string
-	timeStamp string
-	hmac      hash.Hash
-}
-
-// NewSecretsVerifier returns a SecretsVerifier object in exchange for an http.Header object and signing secret
-func NewSecretsVerifier(header http.Header, signingSecret string) (SecretsVerifier, error) {
-	if header["X-Slack-Signature"][0] == "" || header["X-Slack-Request-Timestamp"][0] == "" {
-		return SecretsVerifier{}, errors.New("Headers are empty, cannot create SecretsVerifier")
-	}
-
-	hash := hmac.New(sha256.New, []byte(signingSecret))
-	hash.Write([]byte(fmt.Sprintf("v0:%s:", header["X-Slack-Request-Timestamp"][0])))
-	return SecretsVerifier{
-		slackSig:  header["X-Slack-Signature"][0],
-		timeStamp: header["X-Slack-Request-Timestamp"][0],
-		hmac:      hash,
-	}, nil
-}
-
-func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
-	return v.hmac.Write(body)
-}
-
-// Ensure compares the signature sent from Slack with the actual computed hash to judge validity
-func (v SecretsVerifier) Ensure() error {
-	computed := "v0=" + string(hex.EncodeToString(v.hmac.Sum(nil)))
-	if computed == v.slackSig {
-		return nil
-	}
-
-	return fmt.Errorf("Expected signing signature: %s, but computed: %s", v.slackSig, computed)
-}
+package slack
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// Signature headers
+const (
+	hSignature = "X-Slack-Signature"
+	hTimestamp = "X-Slack-Request-Timestamp"
+)
+
+// SecretsVerifier contains the information needed to verify that the request comes from Slack
+type SecretsVerifier struct {
+	signature string
+	hmac      hash.Hash
+}
+
+func unsafeSignatureVerifier(header http.Header, secret string) (_ SecretsVerifier, err error) {
+	signature := header.Get(hSignature)
+	stimestamp := header.Get(hTimestamp)
+
+	if signature == "" || stimestamp == "" {
+		return SecretsVerifier{}, errors.New("missing headers")
+	}
+
+	hash := hmac.New(sha256.New, []byte(secret))
+	hash.Write([]byte(fmt.Sprintf("v0:%s:", stimestamp)))
+
+	return SecretsVerifier{
+		signature: signature,
+		hmac:      hash,
+	}, nil
+}
+
+// NewSecretsVerifier returns a SecretsVerifier object in exchange for an http.Header object and signing secret
+func NewSecretsVerifier(header http.Header, secret string) (sv SecretsVerifier, err error) {
+	var (
+		timestamp int64
+	)
+
+	stimestamp := header.Get(hTimestamp)
+
+	if sv, err = unsafeSignatureVerifier(header, secret); err != nil {
+		return sv, err
+	}
+
+	if timestamp, err = strconv.ParseInt(stimestamp, 10, 64); err != nil {
+		return SecretsVerifier{}, err
+	}
+
+	diff := absDuration(time.Now().Sub(time.Unix(timestamp, 0)))
+	if diff > 5*time.Minute {
+		return sv, fmt.Errorf("timestamp is too old")
```

Comment:
I'm concerned that this makes it too easy to do the wrong thing. The `SecretsVerifier` returned here will still work, so if the caller doesn't check the error, they may accept a request whose timestamp is too old.

I think the best solution is to move the timestamp parsing/check to `Ensure()`. This simplifies things by letting you get rid of `unsafeSignatureVerifier()`. Alternatively you could change this to return an empty `SecretsVerifier`, but that doesn't feel as nice.

Reply:
I had the same thought, not going to worry about it for now. returning blank SecretVerifier is fine for now.

I was thinking about getting rid of all the error checks in the new method and handling them all in ensure. but decided to pass on that for now. this PR is for the vulnerabilities not the cleaniness =)

Reply:
This code also accepts timestamps with _future_ times in the next 5 mins (due to use of abs).

— Is this intended behaviour?

Reply:
@jimsmart Yes, it's reasonable to allow for clock skew in either direction. The documentation uses abs as well: https://api.slack.com/docs/verifying-requests-from-slack

Reply:
@jimsmart as adam said, either direction is fine as you don't which clock is out of sync in which direction. =)